gcc-5.5 compile error

Compile with gcc-5.5 fails at compiling range.cpp
`[ 64%] Building CXX object core/test/base/CMakeFiles/core_test_base_range.dir/range.cpp.o`

This is part of the error message:
`In file included from /home/u/yanjen/ginkgo/build/third_party/gtest/src/googletest/include/gtest/gtest.h:370:0,
                 from /home/u/yanjen/ginkgo/core/test/base/range.cpp:37:
/home/u/yanjen/ginkgo/core/test/base/range.cpp: In member function ‘virtual void {anonymous}::Range_ComputesBitwiseNot_Test::TestBody()’:
/home/u/yanjen/ginkgo/core/test/base/range.cpp:243:26: error: no match for call to ‘(gko::range<gko::accessor::bitwise_not<{anonymous}::dummy_accessor> >) (int, int, int)’
     EXPECT_EQ(res(1, 2, 3), ~(2 + 6 + 3));
                          ^
In file included from /home/u/yanjen/ginkgo/core/test/base/range.cpp:34:0:
/home/u/yanjen/ginkgo/core/base/range.hpp:331:35: note: candidate: template<class ... DimensionTypes> constexpr decltype (declval<gko::range<Accessor>::accessor>()((forward<DimensionTypes>)(gko::range::operator()::dimensions)...)) gko::range<Accessor>::operator()(DimensionTypes&& ...) const [with DimensionTypes = {DimensionTypes ...}; Accessor = gko::accessor::bitwise_not<{anonymous}::dummy_accessor>]
     GKO_ATTRIBUTES constexpr auto operator()(DimensionTypes &&... dimensions)`

This is strange, for some reason the compiler seems to think that `operand` is not declared:

```
/home/u/mike/git/ginkgo/core/base/range.hpp:599:46: error: ‘operand’ was not declared in this scope
         GKO_ATTRIBUTES static constexpr auto simple_evaluate_impl(     \
                                              ^
/home/u/mike/git/ginkgo/core/base/range.hpp:629:1: note: in expansion of macro ‘GKO_DEFINE_SIMPLE_UNARY_OPERATION’
 GKO_DEFINE_SIMPLE_UNARY_OPERATION(bitwise_not, ~operand);
 ^
```

Since `GKO_DEFINE_SIMPLE_UNARY_OPERATION` is defined as:

```c++
#define GKO_DEFINE_SIMPLE_UNARY_OPERATION(_name, ...)                  \
    struct _name {                                                     \
    private:                                                           \
        template <typename Operand>                                    \
        GKO_ATTRIBUTES static constexpr auto simple_evaluate_impl(     \
            const Operand &operand) -> decltype(__VA_ARGS__)           \
        {                                                              \
            return __VA_ARGS__;                                        \
        }                                                              \
                                                                       \
    public:                                                            \
        template <typename AccessorType, typename... DimensionTypes>   \
        GKO_ATTRIBUTES static constexpr auto evaluate(                 \
            const AccessorType &accessor,                              \
            const DimensionTypes &... dimensions)                      \
            -> decltype(simple_evaluate_impl(accessor(dimensions...))) \
        {                                                              \
            return simple_evaluate_impl(accessor(dimensions...));      \
        }                                                              \
    };
```

The macro substitution `GKO_DEFINE_SIMPLE_UNARY_OPERATION(bitwise_not, ~operand);` should result in:

```c++
struct bitwise_not {
private:
  template <typename Operand>
  GKO_ATTRIBUTES static constexpr auto
  simple_evaluate_impl(const Operand &operand) -> decltype(~operand) {
    return ~operand;
  }
public:
  template <typename AccessorType, typename... DimensionTypes>
  GKO_ATTRIBUTES static constexpr auto
  evaluate(const AccessorType &accessor, const DimensionTypes &... dimensions)
      -> decltype(simple_evaluate_impl(accessor(dimensions...))) {
    return simple_evaluate_impl(accessor(dimensions...));
  }
};
```

which looks perfectly valid to me. Seems like a bug with the older GCC compiler. Anyway, now that you have found a workaround for it, submit a PR, and we'll accept it.

-----------------------------

If you're interested in figuring out what exactly is going on, you may want to try preprocessing 
[this file](https://github.com/ginkgo-project/ginkgo/files/2080035/test.cpp.txt) using `g++ -x c++ -E test.cpp.txt | clang-format` with different GCC versions, and see if they produce different output.

--- a/core/base/range.hpp
+++ b/core/base/range.hpp
@@ -626,7 +626,7 @@ GKO_DEFINE_SIMPLE_UNARY_OPERATION(unary_minus, -operand);
 GKO_DEFINE_SIMPLE_UNARY_OPERATION(logical_not, !operand);
 
 // unary bitwise
-GKO_DEFINE_SIMPLE_UNARY_OPERATION(bitwise_not, ~operand);
+GKO_DEFINE_SIMPLE_UNARY_OPERATION(bitwise_not, ~(operand));
 
 // common functions
 GKO_DEFINE_SIMPLE_UNARY_OPERATION(zero_operation, zero(operand));

